pkg/apis/rbac/types.go

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbac
+
+import (
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+// Authorization is calculated against
+// 1. all deny RoleBinding PolicyRules in the master namespace - short circuit on match
+// 2. all allow RoleBinding PolicyRules in the master namespace - short circuit on match
+// 3. all deny RoleBinding PolicyRules in the namespace - short circuit on match
+// 4. all allow RoleBinding PolicyRules in the namespace - short circuit on match
+// 5. deny by default
+
+const (
+	// PolicyName is the name of Policy
+	PolicyName     = "default"
+	APIGroupAll    = "*"
+	ResourceAll    = "*"
+	VerbAll        = "*"
+	NonResourceAll = "*"
+)
+
+// PolicyRule holds information that describes a policy rule, but does not contain information
+// about who the rule applies to or which namespace the rule applies to.
+type PolicyRule struct {
+	// Verbs is a list of Verbs that apply to ALL the ResourceKinds and AttributeRestrictions contained in this rule.  VerbAll represents all kinds.
+	Verbs sets.String
+	// AttributeRestrictions will vary depending on what the Authorizer/AuthorizationAttributeBuilder pair supports.
+	// If the Authorizer does not recognize how to handle the AttributeRestrictions, the Authorizer should report an error.
+	AttributeRestrictions runtime.Object
+	// APIGroups is the name of the APIGroup that contains the resources.  If this field is empty, then the core API Group is assumed.
+	// If multiple API groups are specified, any action requested against one of the enumerated resources in any API group will be allowed.
+	APIGroups []string
+	// Resources is a list of resources this rule applies to.  ResourceAll represents all resources.
+	Resources sets.String
```

Comment:
You should consider switching these to `[]string` so you don't have to write custom conversion and so that you get stable ordering.  I think the only real problem would be efficiency during evaluation.

Probably something for @erictune and @lavalamp to weigh in on.